crates/storage/src/cache.rs

```diff
@@ -384,4 +391,149 @@ mod tests {
+        let count2 = Arc::clone(&fetch_count);
+        let result2 = cache
+            .get_or_fetch(&key, || {
+                count2.fetch_add(1, Ordering::SeqCst);
+                let (tx, rx) = oneshot::channel();
+                // This should not be called since we have a cache hit
+                tx.send(Err(DbError::Busy)).expect("should not be called");
+                rx
+            })
+            .await;
+
+        assert!(result2.is_ok());
+        assert_eq!(result2.unwrap(), success_value);
+
+        // Critical test: Verify the successful value is still cached and not removed
+        // This validates that the `Arc::ptr_eq` check prevents wrong slot removal
+        let final_count = fetch_count.load(Ordering::SeqCst);
+        assert_eq!(
+            final_count, 1,
+            "Only the first fetch should have been executed"
+        );
+        assert_eq!(cache.get_len(), 1, "Cache should still contain the entry");
+    }
+
+    /// This test specifically validates that error slots are properly removed
+    /// without affecting other cache entries
+    #[tokio::test]
+    async fn test_error_slot_removal() {
+        let cache = CacheTable::<u64, u64>::new(10.try_into().unwrap());
+
+        // Try to fetch a key that will fail
+        let result = cache
+            .get_or_fetch(&42, || {
+                let (tx, rx) = tokio::sync::oneshot::channel();
+                tx.send(Err(DbError::Busy)).expect("send error");
+                rx
+            })
+            .await;
+
+        assert!(result.is_err());
+
+        // The failed slot should be removed from cache
+        assert_eq!(
+            cache.get_len(),
+            0,
+            "Failed slot should be removed from cache"
+        );
+
+        // Now add a successful entry
+        let success_result = cache
+            .get_or_fetch(&42, || {
+                let (tx, rx) = tokio::sync::oneshot::channel();
+                tx.send(Ok(100u64)).expect("send success");
+                rx
+            })
+            .await;
+
+        assert!(success_result.is_ok());
+        assert_eq!(success_result.unwrap(), 100u64);
+        assert_eq!(cache.get_len(), 1, "Successful entry should be cached");
+    }
+
+    #[test]
+    fn test_race_condition_blocking() {
+        let cache = Arc::new(CacheTable::<u64, u64>::new(3.try_into().unwrap()));
+        let fetch_count = Arc::new(AtomicU32::new(0));
+        let key = 99u64;
+        let expected_value = 200u64;
+
+        let cache1 = Arc::clone(&cache);
+        let cache2 = Arc::clone(&cache);
+        let count1 = Arc::clone(&fetch_count);
+        let count2 = Arc::clone(&fetch_count);
+
+        // Spawn two threads that will race to fetch the same key
+        let handle1 = thread::spawn(move || {
+            cache1.get_or_fetch_blocking(&key, || {
+                count1.fetch_add(1, Ordering::SeqCst);
+                thread::sleep(Duration::from_millis(10));
```

Comment:
Is this sleep here to allow for cache access to fail first and then later be populated with this expected value?

Reply:
Just trying to confirm my understanding: if this sleep did not exist, we can't guarantee below that there are 1 error and success count each. Is that right?